### PR TITLE
Add new API model to ACS Engine

### LINKED ISSUE #1019: Add new API model to ACS Engine to enable creating of agent pools only template
**Is this a request for help?**:
No
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
No
---

**What version of acs-engine?**:
n/a
---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes

**What happened**:

**What you expected to happen**:
Create an input API model that ACS Engine will then use to generate ARM template that contains agent pools only.

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by amanohar**
**What this PR does / why we need it**:
This change introduces public API model is needed by ACS Engine to enable creation of agent pool only ARM templates. These templates when deployed will then connect to masters.

**Special notes for your reviewer**:
There are some pending TODOs for validation and more UTs that will come in later.

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/1016)

Fixes #1019

### CODE REVIEW

File: pkg/hostedMasterApi/v20170831/types.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+	Tags     map[string]string     `json:"tags,omitempty"`
+	Type     string                `json:"type,omitempty"`
+
+	Properties *Properties `json:"properties"`
+}
+
+// Properties represents the ACS cluster definition
+type Properties struct {
+	ProvisioningState ProvisioningState   `json:"provisioningState,omitempty"`
+	Version           OrchestratorVersion `json:"version"`
```
Comment (kris-nova) [inline pkg/hostedMasterApi/v20170831/types.go]: Will these be the Kubernetes version?

**Comment by seanknox** [comment]
@amanohar can you tie this back to an existing issue or create one if needed?

File: pkg/hostedMasterApi/v20170831/types.go (hunk 1 above)
Comment (amanohar) [inline pkg/hostedMasterApi/v20170831/types.go]: Yes, this is the Kubernetes version.

**Comment by amanohar** [comment]
Linking to issue: https://github.com/Azure/acs-engine/issues/1019

File: pkg/hostedMasterApi/v20170831/const.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package v20170831
+
+const (
+	// APIVersion is the version of this API
+	APIVersion = "2018-08-31"
```
Comment (kris-nova) [inline pkg/hostedMasterApi/v20170831/const.go]: 2018?

File: pkg/hostedMasterApi/v20170831/types.go (hunk 1 above)
Comment (kris-nova) [inline pkg/hostedMasterApi/v20170831/types.go]: Why aren't we calling it Kubernetes version them?

File: pkg/hostedMasterApi/v20170831/validate.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
+package v20170831
```
Comment (kris-nova) [inline pkg/hostedMasterApi/v20170831/validate.go]: Same question as above? Why are we introducing logic?

File: pkg/hostedMasterApi/v20170831/const.go (hunk 2 above)
Comment (kris-nova) [inline pkg/hostedMasterApi/v20170831/const.go]: Aug 31?

Comment (amanohar) [inline pkg/hostedMasterApi/v20170831/const.go]: Yes

File: pkg/hostedMasterApi/v20170831/types.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+// AccessProfile represents role name and kubeconfig
+type AccessProfile struct {
+	RoleName   string `json:"roleName"`
+	KubeConfig string `json:"kubeConfig"`
+}
+
+// UnmarshalJSON unmarshal json using the default behavior
+// And do fields manipulation, such as populating default value
+func (a *AgentPoolProfile) UnmarshalJSON(b []byte) error {
```
Comment (amanohar) [inline pkg/hostedMasterApi/v20170831/types.go]: Check other examples of how validation is being done for API versions for container service. This is doing the same. At this time I am not planning to do validation in different ways for different API versions or resources types.

File: pkg/hostedMasterApi/v20170831/validate.go (hunk 3 above)
Comment (amanohar) [inline pkg/hostedMasterApi/v20170831/validate.go]: Answered above.

File: pkg/hostedMasterApi/v20170831/types.go (hunk 1 above)
Comment (amanohar) [inline pkg/hostedMasterApi/v20170831/types.go]: Isn't calling it KubernetesVersion redundant (like it's not prepending Kubernetes to other fields)? Plus I thought this was signed off by all in the JSON of the API that I shared last week.

**Pull request merged**